h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -240,69 +266,99 @@ protected MP getModelParams(MP params, Object[] hyperParams) {
-          work *= p.getValue().length;
+      long work = 0;
+      long free_param_combos = 1;
+
+      for (Map<String, Object[]> subspace : _hyperParamSubspaces) {
+        long subspace_param_combos = 1;
+        for (Object[] o : subspace.values()) {
+          subspace_param_combos *= o.length;
         }
+        work += subspace_param_combos;
+      } // work will be zero if there is no subspaces in hyper parameters
+
+      for (Object[] p : _hyperParams.values()) {
+        free_param_combos *= p.length;
       }
+
+      work = work == 0 ? free_param_combos : free_param_combos * work;
+
       return work;
     }
 
+    protected Map<String, Object[]> mergeHashMaps(Map<String, Object[]> hyperparams, Map<String, Object[]> subspaces) {
+      if(subspaces == null) { return hyperparams; }
+      Map<String, Object[]> m = new HashMap<>();
+      m.putAll(hyperparams);
+      m.putAll(subspaces);
+      return m;
+    }
+    
     /** Given a list of indices for the hyperparameter values return an Object[] of the actual values. */
-    protected Object[] hypers(int[] hidx) {
-      Object[] hypers = new Object[_hyperParamNames.length];
+    protected Object[] hypers(Map<String, Object[]> hyperParams, String[] hyperParamNames, int[] hidx) {
+      Object[] hypers = new Object[hyperParamNames.length];
       for (int i = 0; i < hidx.length; i++) {
-        hypers[i] = _hyperParams.get(_hyperParamNames[i])[hidx[i]];
+        hypers[i] = hyperParams.get(hyperParamNames[i])[hidx[i]];
       }
       return hypers;
     }
 
-    protected int integerHash(int[] ar) {
-      Integer[] hashMe = new Integer[ar.length];
+    protected int integerHash(Map<String, Object[]> hyperParams, String[] hyperParamNames, int[] ar, int subspaceNum) {
+      Integer[] hashMe = new Integer[ar.length + 1];
       for (int i = 0; i < ar.length; i++)
-        hashMe[i] = ar[i] * _hyperParams.get(_hyperParamNames[i]).length;
+        hashMe[i] = ar[i] * hyperParams.get(hyperParamNames[i]).length;
+      hashMe[ar.length] = subspaceNum;
       return Arrays.deepHashCode(hashMe);
     }
 
-    private void validateParams() {
+    private void validateParams(Map<String, Object[]> params, boolean isSubspace) {
       // if a parameter is specified in both model parameter and hyper-parameter, this is only allowed if the
       // parameter value is set to be default.  Otherwise, an exception will be thrown.
-      for (String key : _hyperParams.keySet()) {
+      for (String key : params.keySet()) {
         // Throw if the user passed an empty value list:
-        Object[] values = _hyperParams.get(key);
+        Object[] values = params.get(key);
         if (0 == values.length)
           throw new H2OIllegalArgumentException("Grid search hyperparameter value list is empty for hyperparameter: " + key);
 
         if ("seed".equals(key) || "_seed".equals(key)) continue;  // initialized to the wall clock
 
+        if (isSubspace && _hyperParams.get(key) != null) {
```

Comment:
except if you allow null values in your maps and give it a specific meaning, it's usually preferable to use `map.containsKey(key)` to `map.get(key) != null`.
Not extremely important though...